documentation/general/dash-sai-pipeline-packet-flow.md

```diff
@@ -0,0 +1,886 @@
+
+   For example, in the [Inbound from LB](./sdn-pipeline-basic-elements.md#inbound-from-lb) case, the outer encap will start to be mapped to underlay1, the ethernet and IP part of the inner packet will be mapped into underlay0, while the TCP/UDP part will be mapped into overlay. Then, after parsing, we will extract the ethernet and IP parts in underlay0 as overlay, as well as extract the underlay1 as underlay0.
+
+### 5.3. Direction Lookup
+
+In DASH-SAI pipeline, traffic are split into 2 directions: `inbound` and `outbound`. Each direction has its own pipeline (see pipeline overview above). When a new packet arrives, we will assign a direction to the packet, then process the packet in the corresponding pipeline. This ensures us to match the flow and transform the packet in the right way.
+
+### 5.4. Pipeline Lookup
+
+DASH supports multi-tenancy model for traffic handling. A single device can have multiple pipelines, and each pipeline is used to handle traffic for a specific tenant or whichever network function being modeled, such as a load balancer. When a packet arrives, besides direction lookup, we also need pipeline lookup to determine which pipeline to use for processing the packet.
+
+For example, if we like to implement a VM NIC with, then we can model one pipeline as one VM NIC, then use the inner packet MAC to find the pipeline. However, if we would like to implement a load balancer, we can use a DASH-SAI pipeline to represent a load balancer instance and use the Public IP to find the SAI pipeline.
+
+```mermaid
+flowchart TD
+    subgraph ENI
+        EL[ENI Lookup]
+
+        EP0[ENI 0 Pipeline]
+        EP1[ENI 1 Pipeline]
+        EP2[ENI 2 Pipeline]
+    end
+
+    subgraph LoadBalancer
+        VL[Vip Lookup]
+
+        LBP0[LB 0 Pipeline]
+        LBP1[LB 1 Pipeline]
+        LBP2[LB 2 Pipeline]
+    end
+
+    EL --> |ENI 0| EP0
+    EL --> |ENI 1| EP1
+    EL --> |ENI 2| EP2
+
+    VL --> |LB 0| LBP0
+    VL --> |LB 1| LBP1
+    VL --> |LB 2| LBP2
+```
+
+A pipeline can also define its initial matching stages, which will be used for starting processing the packets if no flow is matched.
+
+> NOTE:
+>
+> DASH-SAI pipeline is a logical concept. It doesn't have to be 1 to 1 mapping to a physical ASIC pipeline. The underlying implementation can be as simple as a metadata field update with a specific value, when the entry getting matched, e.g., `ENI = Inner Source/Destination MAC`.
+
+### 5.5. Packet Decap
+
+If a pipeline is found, before processing the packets, all outer encaps will be decap'ed, and with key information saved in metadata bus, such as encap type, source IP and VNI / GRE Key, exposing the inner most packet going through the pipeline. This simplifies the flow matching logic and also allow us to create the reverse flow properly.
```

Comment:
The first sentence is pretty long, is there any way to shorten it?

Reply:
sure thing!